src/api/schema/components/source.rs

```diff
@@ -55,8 +45,21 @@ impl Source {
     pub fn get_component_type(&self) -> &str {
         self.0.component_type.as_str()
     }
-    pub fn get_output_type(&self) -> SourceOutputType {
-        self.0.output_type.into()
+    pub fn get_output_types(&self) -> Vec<SourceOutputType> {
+        let mut t = vec![];
+        self.0
+            .output_type
+            .contains(DataType::Log)
+            .then(|| t.push(SourceOutputType::Log));
+        self.0
+            .output_type
+            .contains(DataType::Metric)
+            .then(|| t.push(SourceOutputType::Metric));
+        self.0
+            .output_type
+            .contains(DataType::Trace)
+            .then(|| t.push(SourceOutputType::Trace));
+        t
     }
```

Comment:
Instead of the repeated logic here, I think this would be a little neater to implement a translation between `DataType` and `SourceOutputType`, and then iterate over the arms:

```rust
impl From<&SourceOutputType> for DataType {
    fn from(s: &SourceOutputType) -> Self {
        match s {
            SourceOutputType::Log => DataType::Log,
            SourceOutputType::Metric => DataType::Metric,
            SourceOutputType::Trace => DataType::Trace,
        }
    }
}

pub fn get_output_types(&self) -> Vec<SourceOutputType> {
    [
        SourceOutputType::Log,
        SourceOutputType::Metric,
        SourceOutputType::Trace,
    ]
    .into_iter()
    .filter(|s| self.0.output_type.contains(s.into()))
    .map(Into::into)
    .collect()
}
```

To save enumerating the options, you could use [strum](https://crates.io/crates/strum), which is used in grok parsing and is a dep of other lib we use.

Reply:
FWIW, I tried the strum implementation just now - adding an [EnumIter](https://docs.rs/strum_macros/0.23.1/strum_macros/derive.EnumIter.html) to `SourceOutputType`:

```rust
#[derive(Debug, Enum, Eq, PartialEq, Copy, Clone, Ord, PartialOrd, EnumIter)]
pub enum SourceOutputType {
    Log,
    Metric,
    Trace,
}
```

And then the method becomes:

```rust
pub fn get_output_types(&self) -> Vec<SourceOutputType> {
    SourceOutputType::iter()
        .filter(|s| self.0.output_type.contains(s.into()))
        .collect()
}
```